### PR TITLE
feat: Add InlineDFG rewrite

### PR DESCRIPTION
**Pull request by acl-cqc**
* The spec doesn't have a lot to say here, but in particular, we handle Order edges both internal (to/from the inner Input/Output) and external (to/from the inlined DFG)
* And, works for permutation (swap) DFGs
* Includes fix to extension inference (search_variable_deps) to handle metas with a solution but no constraints

### CODE REVIEW

**Comment by acl-cqc** [comment]
Test failure is something to do with extension inference ("MismatchedConcreteWithLocations"), haven't figured out what extension annotation I need to add yet

File: src/hugr/rewrite/inline_dfg.rs [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+        }
+        h.remove_node(input).unwrap();
+        h.remove_node(output).unwrap();
+        assert!(h.children(n).next().is_none());
+        h.remove_node(n).unwrap();
+        Ok([n, input, output])
+    }
+
+    fn invalidation_set(&self) -> Self::InvalidationSet<'_> {
+        // TODO should we return Input + Output as well?
```
Comment (cqc-alec) [inline src/hugr/rewrite/inline_dfg.rs:113]: Doesn't seem necessary to me.

File: src/hugr/rewrite/inline_dfg.rs [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+                .is_some_and(|p| h.linked_ports(n, p).next().is_some())
+            {
+                return Err(InlineDFGError::HasOtherEdges(d));
+            };
+        }
+
+        Ok(())
+    }
+
+    fn apply(self, h: &mut impl crate::hugr::HugrMut) -> Result<Self::ApplyResult, Self::Error> {
```
Comment (cqc-alec) [inline src/hugr/rewrite/inline_dfg.rs:58]: Do we expect this to work when the DFG has non-local edges either incoming or outgoing? I can't see why it wouldn't, in either case -- but perhaps worth adding a test.

(Incoming non-local edges from the surrounding DFG would just become local, but I don't think that needs any special handling?)

File: src/extension/infer.rs [hunk 3]
```diff
@@ -652,7 +652,8 @@ impl UnificationContext {
             new_variables = new_variables
                 .into_iter()
                 .filter(|m| seen.insert(*m))
-                .flat_map(|m| self.get_constraints(&m).unwrap())
+                .flat_map(|m| self.get_constraints(&m))
```
Comment (acl-cqc) [inline src/extension/infer.rs:655]: @croyzor this fixes a crash (unwrap of None) here, for a meta that has a solution but no constraints. Does that look ok?

File: src/hugr/rewrite/inline_dfg.rs [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+#[derive(Clone, Debug, PartialEq, Eq, thiserror::Error)]
+pub enum InlineDFGError {
+    /// Node to inline was not a DFG. (E.g. node has been overwritten since the DfgID originated.)
+    #[error("Node {0} was not a DFG")]
+    NotDFG(Node),
+    /// DFG has no parent (is the root)
+    #[error("Node did not have a parent into which to inline")]
+    NoParent,
+    /// DFG has other edges (i.e. Order edges) incoming/outgoing.
+    /// (We don't support such as the new endpoints for such edges is not clear.)
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply src/hugr/rewrite/inline_dfg.rs:23]: Good idea, thanks, will do

File: src/extension/infer.rs (hunk 3 above)
Comment (croyzor) [reply src/extension/infer.rs:655]: I think so 👍

File: src/hugr/rewrite/inline_dfg.rs [hunk 5]
```diff
@@ -70,37 +55,65 @@ impl Rewrite for InlineDFG {
...
-        // Inputs. Just skip any port of the Input node that no out-edges from it.
-        for outp in h.node_outputs(input).collect::<Vec<_>>() {
-            let inport = IncomingPort::from(outp.index());
-            if inport == oth_in {
+        // DFG Inputs. Deal with Order inputs first
+        for (src_n, src_p) in h.linked_outputs(n, oth_in).collect::<Vec<_>>() {
+            // Order edge from src_n to DFG => add order edge to each successor of Input node
+            debug_assert_eq!(Some(src_p), h.get_optype(src_n).other_output_port());
+            for tgt_n in h.output_neighbours(input).collect::<Vec<_>>() {
+                println!("ALAN adding order edge from {src_n} to {tgt_n}");
```
Comment (cqc-alec) [inline src/hugr/rewrite/inline_dfg.rs:63]: Delete this line?

File: src/hugr/rewrite/inline_dfg.rs [hunk 6]
```diff
@@ -278,4 +329,108 @@ mod test {
...
+         *           |  | | .  Cst | NB. Order edge Input to LCst
+         *           |  | |  . |   |
+         *           |  | |   LCst |
+         *           |  |  \ /     |
+         *           |  |  RZ      |
+         *           |  |  |       |
+         *           |  |  meas    |
+         *           |  |  | \     |
+         *           |  |  |  if   |
+         *           |  |  |  .    | NB. State edge if to Output
```
Comment (cqc-alec) [inline src/hugr/rewrite/inline_dfg.rs:350]:
```suggestion
         *           |  |  |  .    | NB. Order edge if to Output
```

**Review by cqc-alec — APPROVED** [review]
Thanks!

**Pull request merged**